src/lib.rs

```diff
@@ -13,14 +13,14 @@ cfg_if::cfg_if! {
         use std::ffi::OsStr;
         use std::os::unix::ffi::OsStrExt;
 
-        fn path_from_bytes(bytes: &[u8]) -> Result<&OsStr, Error> {
-            Ok(OsStr::from_bytes(bytes))
+        fn path_from_bytes(bytes: &[u8]) -> &OsStr {
+            OsStr::from_bytes(bytes)
         }
     } else {
         use std::str;
 
-        fn path_from_bytes(bytes: &[u8]) -> Result<&str, str::Utf8Error> {
-            str::from_utf8(bytes)
+        fn path_from_bytes(bytes: &[u8]) -> &str {
+            unsafe { str::from_utf8_unchecked(bytes) }
```

Comment:
this is wrong. an OsString on unix can be arbitrary nul terminated bytes. it is undefined behaviour to transmute non utf8 bytes to utf8. on windows I think an OsString is a utf16 string.

Reply:
> on windows I think an OsString is a utf16 string.

Windows does use UTF-16 for paths, but with the same qualification that paths are not required to be valid UTF-16. In Rust, however, they are stored as "WTF-8", a UTF-8 superset that allows invalid data.

Reply:
> on windows I think an OsString is a utf16 string

This function also does not claim to be returning an `OsString` only a `&str`. So, if the path was stored in UTF-8 in the binary to begin with, Rust should handle it fine.

I should have access to a Windows machine tonight and I can check to see how `objcopy` stores the gnu-debuglink path in the resulting object (using UTF-16 or UTF-8). If it's UTF-16 (which would very much surprise me) we'll have to think about how to handle this function. If it's UTF-8 we only have to worry about invalid UTF-8, which @luser talked about in our conversation on this pull request.

Reply:
Why can't you leave the code how it is? The existing code correctly handles invalid data, and this isn't performance sensitive. While it is unlikely to be invalid, it is certainly possible.

Reply:
> Why can't you leave the code how it is?

Thinking about it a little more, I'm ok to leave it as is.

My original reasoning was:
I was trying to make the handling of paths on different platforms the same. Previous to this change, `locate_debug_symbols` could return a `PathBuf` on a unix platform, but when run against the same ELF on a Windows platform would fail.

However, I can't think of a situation where a path would be valid on both Windows and unix but not valid UTF-8. So, we can leave this as is.